pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -0,0 +1,552 @@
+		allCIDRs := []string{ciImport.Spec.ServiceCIDR}
+		peerConfigs, err := generatePeerConfigs(allCIDRs, tunnelPeerIP)
+		if err != nil {
+			klog.ErrorS(err, "Parse error for serviceCIDR from remote cluster", "clusterinfoimport", ciImport.Name, "gateway", gw.Name)
+			return err
+		}
+		if c.iAmActiveGW {
+			klog.V(2).InfoS("Adding/Updating flows to remote Gateway Node for Multi-cluster traffic", "clusterinfoimport", ciImport.Name, "cidrs", allCIDRs)
+			localGatewayIP := net.ParseIP(gw.GatewayIP)
+			if err := c.ofClient.InstallMulticlusterGatewayFlows(
+				getCacheKeyWithPrefix(gw.Name+ciImport.Spec.ClusterID),
+				peerConfigs,
+				tunnelPeerIP,
+				localGatewayIP); err != nil {
+				return fmt.Errorf("failed to install flows to remote Gateway in ClusterInfoImport %s: %v", ciImport.Name, err)
+			}
+		} else {
+			klog.V(2).InfoS("Adding/Updating flows to the local active Gateway for Multi-cluster traffic", "clusterinfoimport", ciImport.Name, "cidrs", allCIDRs)
+			tunnelPeerIP := net.ParseIP(gw.InternalIP)
+			if err := c.ofClient.InstallMulticlusterNodeFlows(
+				getCacheKeyWithPrefix(gw.Name+ciImport.Spec.ClusterID),
+				peerConfigs,
+				tunnelPeerIP); err != nil {
+				return fmt.Errorf("failed to install flows to Gateway %s: %v", gw.Name, err)
+			}
+		}
+
+		c.installedCIImports.Add(&ciImportInfo{
+			localGWName: gw.Name,
+			ciImport:    *ciImport,
+		})
+	}
+	c.installedGWName = gw.Name
+	return nil
+}
+
+func (c *MCRouteController) deleteMCNodeFlowsForCIImport(ciImp *mcv1alpha1.ClusterInfoImport, key string) error {
+	if err := c.ofClient.UninstallMulticlusterFlows(key); err != nil {
+		return fmt.Errorf("failed to uninstall multi-cluster flows to remote Gateway Node %s: %v", ciImp.Name, err)
+	}
+	c.installedCIImports.Delete(&ciImportInfo{ciImport: *ciImp})
+	return nil
+}
+
+func (c *MCRouteController) deleteMCNodeFlowsForGateway(gwName string) error {
+	cachedciImps := c.installedCIImports.List()
+	for _, ciImp := range cachedciImps {
+		cached := ciImp.(*ciImportInfo)
+		key := getCacheKeyWithPrefix(gwName + cached.ciImport.Spec.ClusterID)
+		if err := c.ofClient.UninstallMulticlusterFlows(key); err != nil {
+			return fmt.Errorf("failed to uninstall multi-cluster flows to remote Gateway in ClusterInfoImport %s: %v",
+				cached.ciImport.Name, err)
+		}
+		c.installedCIImports.Delete(ciImp)
+	}
+	return nil
+}
+
+func (c *MCRouteController) getAllCIImports(ciImp *mcv1alpha1.ClusterInfoImport) ([]*mcv1alpha1.ClusterInfoImport, error) {
+	var allciImports []*mcv1alpha1.ClusterInfoImport
+	var err error
+	if ciImp == nil {
+		allciImports, err = c.ciImportLister.ClusterInfoImports(c.namespace).List(labels.Everything())
+		if err != nil {
+			return nil, fmt.Errorf("failed to get ClusterInfoImport list, err: %v", err)
+		}
+		if len(allciImports) == 0 {
+			klog.InfoS("No valid remote ClusterInfo imported, do nothing")
+			return nil, nil
+		}
+	} else {
+		allciImports = append(allciImports, ciImp)
```

Comment:
This is confusing. We should not use one func for both cases.

Reply:
Refined.